### PR TITLE
Single container tech stack

### PR DESCRIPTION
**Pull request by akmiller01**

### CODE REVIEW

**Review by nosvalds — CHANGES_REQUESTED** [review]
Overall looks great! I'm not super familiar with the celery/rabbit/elasticsearch so if you want a more critical set of eyes on those maybe @alexlydiate can help

Questions/Feedback:
- So looks like at this point you've got everything (except Postgres as designed) running in one container and a GitHub Workflow that pulls/builds the images, runs the containers, and runs the tests.
- I was able to run the container locally and run the linting successfully, the 'pytest' failed but I'm not super concerned given it passes in the GitHub Actions Workflow.
- I tried to load in the DB dump by running the commands from this [script](https://github.com/IATI/IATI-Internal-Wiki/blob/main/scripts/standard-website-local-sync.sh), but I still get an error rendering the template `In template /usr/src/app/home/templates/home/modules/iati-in-action.html, error at line 13`, were you able to do this successfully? Would be nice to be able to do this for local debugging purposes in the future.
- What are you thinking for deployment options now? I think this probably fits the bill for an Azure Container Instance, so you should be able to steal some of the workflow .yml files from : https://github.com/IATI/azure-docker-aci . The main difference to this is if we want to keep the dev, staging, and www environment setup. Right now that template only supports a develop/production environment. I liked our previous setup in that you could push from a branch to dev. So maybe something like (although we could use this opportunity to re-name things to make them match up):
  - dev.iatistandard.org - can trigger manually from a feature branch to deploy here
  - staging.iatistandard.org - automatically deployed here when code is merged to `develop` branch
  - www.iatistandard.org - automatically deployed here when code is merge to `master` branch?

**Comment by akmiller01** [comment]
> * I tried to load in the DB dump by running the commands from this [script](https://github.com/IATI/IATI-Internal-Wiki/blob/main/scripts/standard-website-local-sync.sh),

Updated the command a bit here https://github.com/IATI/IATI-Internal-Wiki/commit/837b7ed3e468d8a7e66d34a5efa1fcf78e5e32d0

It probably failed if the container was brought up with a different name, or the fact that the public schema is created by `migrate` in the entrypoint.sh.

File: .github/workflows/workflow.yml [hunk 1]
```diff
@@ -14,46 +14,22 @@ on:
...
-          path: ~/.cache/pip
-          key: ${{ runner.os }}-pip-${{ hashFiles('requirements_dev.txt') }}
-          restore-keys: |
-            ${{ runner.os }}-pip-
-            ${{ runner.os }}-
-      - name: Install dependencies
-        run: pip install -r requirements_dev.txt
+      - name: Pull static images
+        run: docker-compose -f docker-compose.dev.yml pull
+      - uses: satackey/action-docker-layer-caching@v0.0.11
```
[the remark this answers is not in the document]
Comment (nosvalds) [reply .github/workflows/workflow.yml]: Looks good!

**Review by nosvalds — COMMENTED** [review]
As for next steps for deployment. I believe you would need to add two main actions/steps:

- Build (using the cache) and push the sha tagged image to the azure container registry 
```yml
env:
  TAG: ${{ github.sha }}
  STAGE: dev
  NAME: aci-iati-website
  DATABASE_NAME: ${{ secrets.DEV_DATABASE_NAME }}
  DB_USERNAME: ${{ secrets.IATI_PG_DEV_USERNAME }}
  DB_PASSWORD: ${{ secrets.IATI_PG_DEV_PASSWORD }}
  DB_HOSTNAME: ${{ secrets.IATI_PG_DEV_HOSTNAME }}
  DB_PORT: ${{ secrets.IATI_PG_PORT }}

      - name: "Build and push image"
        uses: azure/docker-login@v1
        with:
          login-server: ${{ secrets.REGISTRY_LOGIN_SERVER }}
          username: ${{ secrets.REGISTRY_USERNAME }}
          password: ${{ secrets.REGISTRY_PASSWORD }}
      - run: |
          docker build . -t ${{ secrets.REGISTRY_LOGIN_SERVER }}/$STAGE-$NAME:$TAG --cache-from=iati-standard-website_web:latest --build-arg BUILDKIT_INLINE_CACHE=1
          docker push ${{ secrets.REGISTRY_LOGIN_SERVER }}/$STAGE-$NAME:$TAG
```

- Then run the ACI action to set env variable and deploy the image to ACI
```yml
      - name: "Deploy to Azure Container Instances"
        uses: "azure/aci-deploy@v1.1"
        with:
          secure-environment-variables: DATABASE_NAME=${{ env.DATABASE_NAME }} DB_USERNAME=${{ env.DB_USERNAME }} DB_PASSWORD=${{ env.DB_PASSWORD }} DB_HOSTNAME=${{ env.DB_HOSTNAME }} DB_PORT=${{ env.DB_PORT }}
          resource-group: ${{ secrets.RESOURCE_GROUP }}
          dns-name-label: ${{ env.STAGE }}-${{ env.NAME }} # url subdomain: dns-name-label.location.azurecontainer.io/
          image: ${{ secrets.REGISTRY_LOGIN_SERVER }}/${{env.STAGE}}-${{env.NAME}}:${{ env.TAG }}
          registry-login-server: ${{ secrets.REGISTRY_LOGIN_SERVER }}
          registry-username: ${{ secrets.REGISTRY_USERNAME }}
          registry-password: ${{ secrets.REGISTRY_PASSWORD }}
          name: ${{ env.STAGE }}-${{ env.NAME }}
          location: "uksouth"
          ports: 80
            ${{ env.DB_PORT }} # postgres port
```

Might be good for us to chat though about the overall CI/CD plan though. Maybe something like:
- On PR `feature-branch` -> `develop`
  - Builds image
  - Runs tests with docker-compose.dev.yml
  - Pushes image to Azure registry
  - Deploys image to `dev-` ACI instance (only issue here could be if we have multiple PRs going on you would overwrite others work, however this doesn't concern me much for our size team)

- On PR `develop` -> `master`
  - Builds image
  - Runs tests with docker-compose.dev.yml
  - Pushes image to Azure registry
  - Deploys image to `staging-` ACI instance

- On merge to `master`
  - Builds image
  - Runs tests with docker-compose.dev.yml
  - Pushes image to Azure registry
  - Deploys image to `www-` ACI instance

File: .github/workflows/workflow.yml [hunk 2]
```diff
@@ -14,46 +14,31 @@ on:
...
+    env:
+      DOCKER_BUILDKIT: '1'
     steps:
       - uses: actions/checkout@v2
-      - name: Setup Python
-        uses: actions/setup-python@v2
+      - name: Pull static images
+        run: docker-compose -f docker-compose.dev.yml pull
+      - id: cache-docker
+        uses: actions/cache@v1
```
[the remark this answers is not in the document]
Comment (nosvalds) [reply .github/workflows/workflow.yml:24]: @akmiller01 quick fix here.

File: iati/settings/base.py [hunk 3]
```diff
@@ -20,7 +20,7 @@
 BASE_DIR = os.path.dirname(PROJECT_DIR)
 
 ADMINS = (
-    ('Russell Kirkland', 'russell@mashandgravy.co.uk'),
+    ('Alex Miller', 'alex.miller@devinit.org'),
```
Comment (nosvalds) [inline iati/settings/base.py:23]: Would it also be worth making code@iatistandard.org an admin?

**Comment by nosvalds** [comment]
Looking great, just have the 2 comments inline. Great to see the CI Build is now ~10min instead of 35min!

What are you thinking for next steps? I'm thinking we probably want to test the deployment to Azure ACI from this feature branch rather than merging to `develop`.

**Comment by akmiller01** [comment]
> What are you thinking for next steps? I'm thinking we probably want to test the deployment to Azure ACI from this feature branch rather than merging to `develop`.

Yes, agreed. Set up the pre-requisite Azure blob storage, Postgres database. I know the blob container needs to exist before deploy, the Azure storages app explicitly said it won't create the container itself.

And then we can change the `if` statement to let it try and deploy from here.

**Review by nosvalds — APPROVED** [review]

**Review by alexlydiate — APPROVED** [review]

**Pull request merged**